driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,197 @@
+use super::manually_updated_price_source::ManuallyUpdatedPriceSource;
+use super::{price_source::PriceSource, Token};
+use crate::models::TokenId;
+use anyhow::Result;
+use std::collections::HashMap;
+use std::sync::{
+    mpsc::{self, RecvTimeoutError, Sender},
+    Arc, Mutex,
+};
+use std::thread::{self, JoinHandle};
+use std::time::{Duration, Instant};
+
+/// Implements `PriceSource` in a non blocking way by updating prices in a
+/// thread and reusing previous results.
+pub struct ThreadedPriceSource {
+    // Shared between this struct and the thread. The background thread writes
+    // the prices, the thread calling `get_prices` reads them.
+    non_blocking_price_source: Arc<Mutex<ManuallyUpdatedPriceSource>>,
+    // Channel over which the thread can be told to update the prices.
+    channel_to_thread: Sender<ThreadMessage>,
+}
+
+impl ThreadedPriceSource {
+    /// The join handle represents the background thread. It can be used to
+    /// verify that it exits when the created struct is dropped.
+    #[allow(dead_code)]
+    pub fn new<T: 'static + PriceSource + Send>(
+        price_source: T,
+        update_interval: Duration,
+    ) -> (Self, JoinHandle<()>) {
+        let non_blocking_price_source = Arc::new(Mutex::new(ManuallyUpdatedPriceSource::new()));
+        let (sender, receiver) = mpsc::channel::<ThreadMessage>();
+
+        let nbps = non_blocking_price_source.clone();
+        // vk: Clippy notes that the following loop and match can be written as
+        // a `while let` loop but I find it clearer to make all cases explicit.
+        #[allow(clippy::while_let_loop)]
+        let join_handle = thread::spawn(move || loop {
+            match receiver.recv_timeout(update_interval) {
+                Ok(ThreadMessage::UpdateImmediately) | Err(RecvTimeoutError::Timeout) => {
+                    let now = Instant::now();
+                    let cutoff = now.checked_sub(update_interval).unwrap_or(now);
```

Comment:
Even if we update immediately we only update the tokens that were not updated within the update interval? I would have expected that in that case we update all tokens.